File: leetcode_coding_challang/two_pointer_approach_medium/881_boats_to_save_people/boats_to_save_people.java
import java.util.Arrays;

class Solution {
    /**
     * Determines minimum boats needed to rescue all people
     * 
     * @param people Array of people's weights
     * @param limit Maximum weight capacity per boat
     * @return Minimum number of boats required
     */
    public int numRescueBoats(int[] people, int limit) {
        // Step 1: Sort the array to enable two-pointer approach
        // Time complexity: O(n log n) where n is number of people
        Arrays.sort(people);
        
        // Step 2: Initialize pointers at array boundaries
        int left = 0;                    // Lightest person index
        int right = people.length - 1;   // Heaviest person index
        int boats = 0;                   // Boat counter
        
        // Step 3: Greedily pair people from extremes
        // Continue until all people are assigned
        while (left <= right) {
            // Check if lightest and heaviest can share a boat
            if (people[left] + people[right] <= limit) {
                // They can share - move left pointer forward
                // This person is now assigned to a boat
                left++;
            }
            
            // The heaviest person always gets assigned in this iteration
            // Either paired with lightest (if condition true) or alone
            right--;
            boats++;
        }
        
        return boats;
    }
}

File: leetcode_coding_challang/two_pointer_approach_medium/881_boats_to_save_people/boats_to_save_people.js
/**
 * @param {number[]} people - array of people's weights
 * @param {number} limit - maximum weight capacity per boat
 * @return {number} - minimum number of boats needed
 */
function numRescueBoats(people, limit) {
  // Step 1: Sort people by weight (ascending order)
  // This allows us to efficiently access lightest and heaviest people
  people.sort((a, b) => a - b);

  // Step 2: Initialize two pointers at the extremes
  let left = 0; // Points to lightest person
  let right = people.length - 1; // Points to heaviest person
  let boats = 0; // Counter for boats needed

  // Step 3: Process people from both ends toward the middle
  while (left <= right) {
    // Try to pair the heaviest with the lightest
    if (people[left] + people[right] <= limit) {
      // Success! They can share a boat
      // Move left pointer right (lightest person is assigned)
      left++;
    }
    // Whether paired or not, the heaviest person gets on a boat
    // If paired, both are assigned; if not, heavy person goes alone
    right--;
    boats++;
  }

  return boats;
}

// Example usage with detailed trace:
console.log(numRescueBoats([3, 2, 2, 1], 3)); // Output: 3
// Trace: sorted = [1,2,2,3]
// Iteration 1: 1+3=4 > 3, send 3 alone, boats=1, right=2
// Iteration 2: 1+2=3 ≤ 3, pair them, boats=2, left=1, right=1
// Iteration 3: 2 alone (left==right), boats=3, right=0
